happytransformer/fine_tuning_util.py

```diff
@@ -0,0 +1,78 @@
+"""
+Contains functions that are shared amongst the children of the HappyTrainer class.
+
+Based on
+https://github.com/huggingface/transformers/blob/master/examples/pytorch/language-modeling/run_mlm.py
+and
+https://github.com/huggingface/transformers/blob/master/examples/pytorch/language-modeling/run_clm.py
+"""
+
+
+def preprocess_concatenate(tokenizer, dataset, preprocessing_processes, mlm=True):
+    """
+    :param tokenizer: tokenizer for a transformer model
+    :param dataset: a datasets.Dataset object
+    :param preprocessing_processes:
+    :param mlm:
+
+    :return:
+    """
+
+    max_input_length = tokenizer.model_max_length
+
+
+    def tokenize_function(example):
+        return tokenizer(example["text"],
+                         add_special_tokens=True, truncation=True,)
+
+    tokenized_dataset = dataset.map(tokenize_function, batched=True,
+                                      num_proc=preprocessing_processes,
+                                      remove_columns=["text"])
+
+    def group_texts(examples):
+        concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
+        total_length = len(concatenated_examples[list(examples.keys())[0]])
+        output_length = (total_length // max_input_length) * max_input_length
+
+        # if  total_length is less than the max_input_length length
+        # then it causes an error unless the code below is used.
+        # this is due to total_length being truncated to 0
+        if output_length == 0:
+            output_length = total_length
```

Comment:
What is the desired `output_length` when `total_length==max_input_length+1`?

Reply:
example: 
max_input_length =  512
total_length = 513
output_length = (total_length // max_input_length) * max_input_length = 1 * 512 = 512

By adding lines 40 and 41, it guarantees that at least some training will occur even if the length of the  input is less than the max_input_length.

Reply:
So you're okay with discarding the end of the concatenated examples?

Reply:
That's how it's done in the official documentation. I did some experimenting, and if you add the remainder to to the end of the input_ids as a list that's shorter than the other lists, then it causes an error. All inputs must be the same length. 

The official documentation mentions that you can add padding to prevent. It also mentioned that each model type has a different padding strategy.  I think we could add this functionality later for popular model types like GPT-2 in the future. 

Also, the tokenizer function no longer truncates. In the official documentation  only the line-by-line method requires the tokenizer to truncate which makes sense.

Reply:
Okay can we at least drop in a `#TODO` to replace the truncation with padding then?

Reply:
I'll create an issue